avro-codegen/src/main/java/com/linkedin/avroutil1/codegen/SpecificRecordClassGenerator.java

```diff
@@ -1628,6 +1696,89 @@ private void addGetByIndexMethod(TypeSpec.Builder classBuilder, AvroRecordSchema
+
+      if(fieldClass != null ) {
+        if (fieldClass.equals(long.class)) {
+          // If field is of type long, add method with int parameter
+          // int value can safely be cast to long
+          numberSetter
+              .addParameter( int.class, escapedFieldName)
+              .addStatement("this.$1L = (long) $1L", escapedFieldName);
+        } else if (fieldClass.equals(int.class)) {
+          // If field is of type int, add method with long parameter
+          // This method will error if the long input is greater than Integer.MAX_VALUE
+
+          CodeBlock castToInt = CodeBlock
+              .builder()
+              .beginControlFlow("if ($1L <= Integer.MAX_VALUE && $1L >= Integer.MIN_VALUE)", escapedFieldName)
+              .addStatement("this.$1L = (int) $1L", escapedFieldName)
+              .endControlFlow()
+              .beginControlFlow("else")
+              .addStatement("throw new org.apache.avro.AvroRuntimeException(\"long value cannot be cast to int\")")
+              .endControlFlow()
+              .build();
+
+          numberSetter
+              .addParameter(long.class, escapedFieldName)
+              .addCode(castToInt);
+        }
+
+      } else if (field.getSchema() != null && field.getSchema().type().equals(AvroType.UNION)) {
+        TypeName typeName = SpecificRecordGeneratorUtil.getTypeName(field.getSchemaOrRef().getSchema(),
+            field.getSchemaOrRef().getSchema().type(), true,
+            config.getDefaultMethodStringRepresentation());
+
+        CodeBlock nullCheck = CodeBlock.builder()
+            .beginControlFlow("if ($1L == null)", escapedFieldName)
+            .addStatement("this.$1L = null", escapedFieldName)
+            .endControlFlow()
+            .build();
+
+        if (typeName.equals(ClassName.get(Long.class))) {
+          numberSetter
+              .addParameter(ClassName.get(Integer.class), escapedFieldName)
+              .addCode(nullCheck)
+              .beginControlFlow("else")
+              .addStatement("this.$1L = Long.valueOf($1L)", escapedFieldName)
+              .endControlFlow();
+
+        } else if (typeName.equals(ClassName.get(Integer.class))) {
+
+          CodeBlock castToInt = CodeBlock
+              .builder()
+              .beginControlFlow("else if ($1L <= Integer.MAX_VALUE && $1L >= Integer.MIN_VALUE)", escapedFieldName)
+              .addStatement("this.$1L = $1L.intValue()", escapedFieldName)
+              .endControlFlow()
+              .beginControlFlow("else")
+              .addStatement("throw new org.apache.avro.AvroRuntimeException(\"Long value cannot be cast to Integer\")")
+              .endControlFlow()
+              .build();
+
+          numberSetter
+              .addParameter(ClassName.get(Long.class), escapedFieldName)
+              .addCode(nullCheck)
+              .addCode(castToInt);
+        }
+      }
+    }
+    return numberSetter == null ? null : numberSetter.build();
+  }
```

Comment:
Not a change in this PR but I noticed that the deepcopy method has "String" as type. Is that correct?